### ISSUE TITLE
Trying to implement prelu and multi dim softmax, need a code check

### ISSUE THREAD

**Issue by seranus** [chaosmail/deeplearnjs-caffe#9]
I'm trying to port mtcnn to deeplearnjs and I'm missing prelu and in tensorflow implementation uses a multi div softmax. I gave it a try but I'm not sure what I'm doing. I'd appreciate if someone could check it

mlti div soft max python

https://github.com/tensorflow/tensorflow/issues/210

    def softmax(self, target, axis, name=None):
        max_axis = tf.reduce_max(target, axis, keep_dims=True)
        target_exp = tf.exp(target-max_axis)
        normalize = tf.reduce_sum(target_exp, axis, keep_dims=True)
        softmax = tf.div(target_exp, normalize, name)
        return softmax`

converted

          case 'softmax':
            const softmaxParam = caffe.SoftmaxParameter.create(layer.softmaxParam);
            const axis = softmaxParam.axis;
            const inp = input as Array3D;
            const maxAxis = math.max(inp, axis, true);
            const inpSub = math.sub(inp, maxAxis);
            const targetExp = math.exp(inpSub);
            const normalize = math.sum(targetExp, axis, true);
            const softMax = math.divide(targetExp, normalize);
            return softMax;`

prelu
        `case 'prelu':

            const alpha = blobs[0].as1D() as Array1D;
            return math.prelu(input as NDArray, alpha);`

**Comment by chaosmail** [comment]
Hi @seranus 

thanks for your feedback. The underlying implementation of `softmax` in deeplearn.js uses the exact method as in tensorflow (you can find the source here: https://github.com/PAIR-code/deeplearnjs/blob/42a2c34869b7693759a5dd329da8cc6aaed941d8/src/ops/softmax.ts#L45), which is numerically more stable than applying exp to every value before the division.

The prelu example might be a perfect opportunity to send a PR on this repo, don't you think? Let me know if you need help with this.

Best, Christoph

**Comment by seranus** [comment]
your right, I didn't notice softmax had optional params, so softmax would be just

      case 'softmax':
        const softmaxParam = caffe.SoftmaxParameter.create(layer.softmaxParam);
        const axis = softmaxParam.axis;

        return math.softmax(input as NDArray, axis);`

and prelu

`  alphas = tf.get_variable('alpha', _x.get_shape()[-1],
                       initializer=tf.constant_initializer(0.0),
                        dtype=tf.float32)`

can be converted to `const alpha = blobs[0].as1D() as Array1D;` ?
